docs; adds *tidyselect*-like DSL

strengejacke opened the pull request.

IndrajeetPatil commented on R/data_relocate.R:
```diff
@@ -10,6 +11,14 @@
 #'
+#' @inherit data_rename seealso
+#'
+#' @details `data_relocate()` will reorder columns to specific positions,
```
Super minor: Maybe better to have one line per individual function? Given how similar the function names, it's easy to miss the change in function name in the second sentence.

strengejacke replied:
Do you mean separating these two sentences with a new line?

IndrajeetPatil replied:
Yes, like so:

```r
#' @details 
#'
#' `data_relocate()` will reorder columns to specific positions, indicated by `before` or `after`. 
#'
#' `data_reorder()` will instead always move selected columns to the beginning of the data frame .
```

IndrajeetPatil commented on R/data_findcols.R:
```diff
@@ -1,10 +1,16 @@
-#' @rdname data_reorder
+#' @param pattern A regular expression (as character string), representing the
+#'   pattern to be matched in the in column names.
+#' @param starts_with,end_with Character string, containing the string to be
```
Should be `ends_with` instead of  `end_with`.

strengejacke commented on R/data_remove.R:
```diff
@@ -1,10 +1,10 @@
 #' @export
 data_remove <- function(data, pattern, ...) {
-  new <- data[!names(data) %in% pattern]
+  new <- data[!grepl(pattern, names(data))]
```
I think this makes more sense, but not sure where this function is used in our packages? It might break code. We could also add an argument like `regex`  or `fixed`. to switch whether `pattern` is a regular expression or a fixed string.

@DominiqueMakowski @IndrajeetPatil ?

IndrajeetPatil replied:
> not sure where this function is used in our packages?

In multiple places:
https://github.com/search?q=org%3Aeasystats+data_remove&type=code

Not sure how this change affects them.

strengejacke replied:
ok, we need to take into account multiple patterns.

strengejacke commented:
Ok, I added some NSE stuff to `data_findcols()` and `data_remove()`, which should also work when called from outside functions:

``` r
library(datawizard)
data_findcols(iris, pattern = starts_with("Sepal"))
#> [1] "Sepal.Length" "Sepal.Width"

data_findcols(iris, pattern = ends_with("Width"))
#> [1] "Sepal.Width" "Petal.Width"

data_findcols(iris, pattern = "\\.")
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

data_findcols(iris, pattern = contains("Wid"))
#> [1] "Sepal.Width" "Petal.Width"

data_findcols(iris, c("Petal.Width", "Sepal.Length"))
#> [1] "Petal.Width"  "Sepal.Length"

data_findcols(mtcars, c("am", "gear", "cyl"))
#> [1] "am"   "gear" "cyl"

data_findcols(mtcars, c("vam", "gear", "cyl"))
#> [1] "gear" "cyl"

data_findcols(iris, Spec)
#> [1] "Species"

data_findcols(mtcars, ends_with = "abc")
#> character(0)

data_findcols(mtcars, "rb$")
#> [1] "carb"

data_findcols(mtcars, "^c")
#> [1] "cyl"  "carb"

test_fun <- function(data, i) {
  data_findcols(data, pattern = i)
}
test_fun(iris, "Sep")
#> [1] "Sepal.Length" "Sepal.Width"
x <- "Pet"
test_fun(iris, x)
#> [1] "Petal.Length" "Petal.Width"
```

<sup>Created on 2022-03-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

bwiernik commented:
Are `contains()`, etc. the tidyselect functions?

strengejacke commented:
> Are `contains()`, etc. the tidyselect functions?

no, internals. see https://github.com/easystats/datawizard/blob/ab6e389e6f319c1ac4ec1707072ac2fda83d2074/R/select_helpers.R

strengejacke commented:
@IndrajeetPatil @mattansb @bwiernik @DominiqueMakowski 

This PR actually has two parts. First, I revised the docs for those cases where arguments had same names, but slightly different contexts, or where arguments were not described in details. Second, I added some kind of NSE and select-helper stuff to some functions.

This is currently for `data_remove()`, `data_extract()` and `data_findcols()`. What do you say?

strengejacke commented:
`data_extract()` serves both as `select()` and `pull()` (this was possible before, but now even more convenient due to select helpers).

strengejacke commented:
I have added some features to `data_extract()`. This function can extract (and return) either a vector, or multiple variables as data frame. That behaviour existed before. However, to make that function type-safe, I added some arguments that let you control what to return, and revised the docs accordingly.

mattansb commented:
> > Are contains(), etc. the tidyselect functions?
>
> no, internals. 

I think we should avoid masking / conflicting with `tidyverse` - I use datawizrd with dplyr all the time, and I can already see people on SO: "why won't dplyr work???" "Oh, this is a known issue, don't load MASS or datawizard with dplyr".

IndrajeetPatil commented:
I strongly agree with Mattan here. We shouldn't be masking any of _tidyselect_ function.

I think a good compromise is to use similar names, but in a different tense. The  _tidyselect_ functions use simple present tense, while  _datawizard_ functions can use continuous present tense. 

For example,

`contains` -> `containing`
`matches` -> `matching`
`starts_with` -> `starting_with`
`ends_with` -> `ending_with`
...
etc.

This way, 

- users familiar with _tidyverse_ don't need to know completely new function naming pattern
- auto-completion will reveal the similarity between these function families across packages

strengejacke commented:
There is no conflict, because we have no such functions. Is just a evaluation of strings containing the function names, and then create an appropriate reprex string.

strengejacke commented:
``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tidyselect)
library(datawizard)

colnames(data_extract(mtcars, contains("a")))
#> [1] "drat" "am"   "gear" "carb"
colnames(select(mtcars, contains("a")))
#> [1] "drat" "am"   "gear" "carb"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

strengejacke commented:
``` r
library(datawizard)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tidyselect)

colnames(data_extract(mtcars, contains("a")))
#> [1] "drat" "am"   "gear" "carb"
colnames(select(mtcars, contains("a")))
#> [1] "drat" "am"   "gear" "carb"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

IndrajeetPatil commented:
Sorry, should have looked at the implementation before commenting. 

Just looked at the `.evaluate_pattern()` function and I understand now how this is working.

Really awesome and cool stuff. Thanks, Daniel!

strengejacke commented:
We could make use of the autocompletion when dplyr is loaded and accept additional arguments, however, my regex knowledge is too limited to parse this correctly:


``` r
gsub("contains\\(\"(.*)\"\\)", "\\1", "contains(\"a\", ignore.case = TRUE)")
```

But we could also simply have an additional argument `ignore_case`, so the select helpers only accept a character string w/o any additional arguments.

strengejacke commented:
Forgot an important one: from-to:

``` r
colnames(datawizard::data_extract(mtcars, cyl:drat))
#> [1] "cyl"  "disp" "hp"   "drat"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

mattansb commented:
Oh, that is VERY COOL!

strengejacke commented:
Not sure, though, why the two tests are falling now?

strengejacke commented:
damn, there's a conflict with _poorman_, because it handles select-helpers differently:

``` r
colnames(datawizard::data_remove(iris, contains("Sep")))
#> [1] "Petal.Length" "Petal.Width"  "Species"

library(poorman)
#> 
#>   I'd seen my father. He was a poor man, and I watched him do astonishing things.
#>     - Sidney Poitier
#> 
#> Attaching package: 'poorman'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
colnames(datawizard::data_remove(iris, contains("Sep")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

strengejacke commented:
``` r
colnames(datawizard::data_remove(iris, contains("Sep")))
#> [1] "Petal.Length" "Petal.Width"  "Species"

library(poorman)
#> 
#>   I'd seen my father. He was a poor man, and I watched him do astonishing things.
#>     - Sidney Poitier
#> 
#> Attaching package: 'poorman'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
colnames(datawizard::data_remove(iris, contains("Sep")))
#> [1] "Petal.Length" "Petal.Width"  "Species"


colnames(datawizard::data_remove(iris, contains("Sep")))
#> [1] "Petal.Length" "Petal.Width"  "Species"


mtcars |> poorman::count(cyl)
#>   cyl  n
#> 1   4 11
#> 2   6  7
#> 3   8 14

colnames(datawizard::data_remove(iris, contains("Sep")))
#> [1] "Petal.Length" "Petal.Width"  "Species"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

strengejacke commented:
The last commit add select-helpers to `data_long()`. No tests added yet.

strengejacke commented:
We should add a vignette on select-helpers ;-)
 But for now, I'll stop here. Any more functions in _datawizard_ that need select helpers? We could open another PR for that after merging this one.

strengejacke commented:
We have some inconsistencies in the default behaviours, to not break other packages:

- `data_extract()` has no `fixed` argument, but internally defaults it to `TRUE` (https://github.com/easystats/datawizard/blob/e87baf6681d31404843418f07da4f1f1baf2b825/R/data_extract.R#L98)
- `data_findcols()` has no `fixed` argument and doesn't need it, because it _always_ uses `grepl()` (https://github.com/easystats/datawizard/blob/e87baf6681d31404843418f07da4f1f1baf2b825/R/data_findcols.R)
- `data_remove()` has a `fixed` argument, default `TRUE`, because it used to expect a character vector only(https://github.com/easystats/datawizard/blob/e87baf6681d31404843418f07da4f1f1baf2b825/R/data_remove.R)
- `reshape_longer()` has no `fixed` argument, but internally needs to default to `FALSE` (https://github.com/easystats/datawizard/blob/e87baf6681d31404843418f07da4f1f1baf2b825/R/data_reshape.R#L97).

strengejacke commented:
I hopefully covered most/all(?) use cases, and everything *should* work when called from other functions:

``` r
library(datawizard)
colnames(data_remove(iris, "Sepal"))
#> Warning in .evaluated_pattern_to_colnames(pattern, data, ignore_case, verbose):
#> Following variable(s) were not found: Sepal
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, starts_with("Sepal")))
#> [1] "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, "Species"))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

colnames(data_remove(iris, "species"))
#> Warning in .evaluated_pattern_to_colnames(pattern, data, ignore_case, verbose):
#> Following variable(s) were not found: species
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, "species", ignore_case = TRUE))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

colnames(data_remove(iris, species))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, Species))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

colnames(data_remove(iris, species, ignore_case = TRUE))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

colnames(data_remove(iris, c("Sepal.Length", "Sepal.Width")))
#> [1] "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, Sepal.Width:Petal.Width))
#> [1] "Sepal.Length" "Species"

colnames(data_remove(iris, contains("Sep")))
#> [1] "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, contains("sep")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, contains("sep"), ignore_case = TRUE))
#> [1] "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, 1:3))
#> [1] "Petal.Width" "Species"

colnames(data_remove(iris, regex("\\.")))
#> [1] "Species"

colnames(data_remove(iris, regex("^sep")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(data_remove(iris, regex("^sep"), ignore_case = TRUE))
#> [1] "Petal.Length" "Petal.Width"  "Species"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

strengejacke commented:
Called from functions, select-helpers won't work, and only quoted names or numeric indices

``` r
library(datawizard)

testit <- function(data, select, ...) {
  data_remove(data, select, ...)
}

colnames(testit(iris, "Sepal"))
#> Warning: Following variable(s) were not found: Sepal
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, starts_with("Sepal")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(testit(iris, "Species"))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

colnames(testit(iris, "species"))
#> Warning: Following variable(s) were not found: species
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(testit(iris, "species", ignore_case = TRUE))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"

# bare names won't work
colnames(testit(iris, species))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# bare names won't work
colnames(testit(iris, Species))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# bare names won't work
colnames(testit(iris, species, ignore_case = TRUE))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(testit(iris, c("Sepal.Length", "Sepal.Width")))
#> [1] "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, Sepal.Width:Petal.Width))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, contains("Sep")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, contains("sep")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, contains("sep"), ignore_case = TRUE))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

colnames(testit(iris, 1:3))
#> [1] "Petal.Width" "Species"

# select helper won't work
colnames(testit(iris, regex("\\.")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, regex("^sep")))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"

# select helper won't work
colnames(testit(iris, regex("^sep"), ignore_case = TRUE))
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"
```

<sup>Created on 2022-03-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

DominiqueMakowski reviewed (approved):
amazing

strengejacke commented:
Ok, I'll merge this, and we can improve docs or other functions later, when necessary. Select-helpers don't currently work when called from other functions, but this a) doesn't break any existing code and b) is not top-priority.

IndrajeetPatil commented:
Thanks a lot, Daniel, for this _tour de force_ PR.

We also need to make sure that we log these changes in `NEWS.md`.
